editor.planx.uk/src/@planx/graph/index.ts

```diff
@@ -504,87 +504,76 @@ export const sortIdsDepthFirst =
+      output.push(`Removed ${node.type ? TYPES[node.type] : "node"} from branch`);
+    }
+  };
+
   ops.map((op) => {
     const node = graph[op.p?.[0]];
-    // Updating an object or its properties (update = delete + insert)
-    if (Object.keys(op).includes("od") && Object.keys(op).includes("oi")) {
-      if (op.od.type && op.oi.type) {
-        output.push(
-          `Replaced ${TYPES[op.od.type]} "${
-            op.od.data?.title || op.od.data?.text
-          }" with ${TYPES[op.oi.type]} "${
-            op.oi.data?.title || op.oi.data?.text
-          }"`,
-        );
-      } else if (op.p.includes("data")) {
-        if (node.type) {
-          output.push(
-            `Updated ${TYPES[node.type]} ${op.p?.[2]} from "${op.od}" to "${
-              op.oi
-            }"`,
-          );
-        } else {
-          output.push(
-            `Updated node ${op.p?.[2]} from "${op.od}" to "${op.oi}"`,
-          );
-        }
-      } else if (op.p.includes("edges")) {
-        output.push(
-          `Updated order of ${node.type ? TYPES[node.type] : "graph"} edges`,
-        );
-      }
-      // Adding (inserting) an object or its properties
-    } else if (Object.keys(op).includes("oi")) {
-      if (op.oi.type) {
-        output.push(
-          `Added ${TYPES[op.oi.type]} "${
-            op.oi.data?.title || op.oi.data?.text
-          }"`,
-        );
-      } else if (op.p.includes("data")) {
-        if (node.type) {
-          output.push(`Added ${TYPES[node.type]} ${op.p?.[2]} "${op.oi}"`);
-        } else {
-          output.push(`Added node ${op.p?.[2]} "${op.oi}"`);
-        }
-      } else if (op.p.includes("edges")) {
-        if (node.type) {
-          output.push(`Added ${TYPES[node.type]} to branch`);
-        } else {
-          output.push(`Added node to branch`);
-        }
-      }
-      // Deleting an object or its properties
-    } else if (Object.keys(op).includes("od")) {
-      if (op.od.type) {
-        output.push(
-          `Deleted ${TYPES[op.od.type]} "${
-            op.od.data?.title || op.od.data?.text
-          }"`,
-        );
-      } else if (op.p.includes("data")) {
-        if (node.type) {
-          output.push(`Deleted ${TYPES[node.type]} ${op.p?.[2]} "${op.od}"`);
-        } else {
-          output.push(`Deleted node ${op.p?.[2]} "${op.od}"`);
-        }
-      } else if (op.p.includes("edges")) {
-        if (node.type) {
-          output.push(`Deleted ${TYPES[node.type]} from branch`);
-        } else {
-          output.push(`Deleted node from branch`);
-        }
-      }
-      // Updating the _root list (update = list insert or list delete)
-    } else if (
-      Object.keys(op).includes("li") &&
-      Object.keys(op).includes("ld")
-    ) {
-      if (op.p.includes("edges") && op.p.includes("_root")) {
-        output.push(`Re-ordered the graph`);
-      }
+    const operationTypes = Object.keys(op);
+
+    if (operationTypes.includes("od") && operationTypes.includes("oi")) {
```

Comment:
nit: If we made this a type guard, we could avoid the casting here if we wanted to - 

```ts
const isUpdate = (op: OT.Op): op is OT.Object.Replace => {
  const operationTypes = Object.keys(op);
  return operationTypes.includes("od") && operationTypes.includes("oi");
}
```

Reply:
This is super useful to keep in mind & I'll definitley come back to it - will personally feel more confident full type-guarding after more testing of various operation scenarios !